### PR TITLE
[server] Improve observability regarding communication with messagebus

### LINKED ISSUE #3578: [server] Improve observability regarding communication with messagebus
`server` is one of the protagonists of the outages in the past few days. One thing that we're not satisfied with is that we don't have enough observability about its communication with messagebus.

Following up today's post-mortem meeting, we agreed on a couple of improvements in that area:
* Log an error message when `server` tries to create a topic and fails
* Create a new metric `gitpod_server_topic_read_total` (Counter) that increases every time `server` tries to read from a topic.

### PR DESCRIPTION
**Pull request by ArthurSens**
When ready, this PR will include a counter metric that increases every time server tries to read a topic from messagebus and also log when it fails to do so.

Fixes #3578

### CODE REVIEW

**Comment by ArthurSens** [comment]
@geropl @csweichel The metric should be ready to review. 

Regarding the log message, different from Go, typescript functions don't seem to return an error when something fails. Any ideas on how to log an error during listening then? From what I could understand, the error occurs at messagebus' code instead of server's

File: components/server/src/workspace/messagebus-integration.ts [hunk 1]
```diff
@@ -117,20 +118,23 @@ export class MessageBusIntegration extends AbstractMessageBusIntegration {
         const listener = new HeadlessWorkspaceLogListener(this.messageBusHelper, callback, workspaceID);
         const cancellationTokenSource = new CancellationTokenSource()
         this.listen(listener, cancellationTokenSource.token);
+        increaseMessagebusTopicReads('headless_workspace_log')
```
Comment (csweichel) [inline components/server/src/workspace/messagebus-integration.ts:121]: this should be the actual topic we're listening on ... same goes for the other two

File: components/server/src/workspace/messagebus-integration.ts [hunk 2]
```diff
@@ -117,6 +118,7 @@ export class MessageBusIntegration extends AbstractMessageBusIntegration {
         const listener = new HeadlessWorkspaceLogListener(this.messageBusHelper, callback, workspaceID);
         const cancellationTokenSource = new CancellationTokenSource()
         this.listen(listener, cancellationTokenSource.token);
+        increaseMessagebusTopicReads(listener.topic.name)
         return Disposable.create(() => cancellationTokenSource.cancel())
     }
```
Comment (geropl) [inline components/server/src/workspace/messagebus-integration.ts:124]: Why not listen in `listenForPrebuildUpdatableQueue` as well? :thinking:

**Comment by geropl** [comment]
The metrics currently emitted:
`gitpod_server_topic_reads_total{topic="topic"} 5`

File: components/server/src/workspace/messagebus-integration.ts (hunk 2 above)
Comment (ArthurSens) [reply components/server/src/workspace/messagebus-integration.ts:124]: PrebuildUpdatableQueue's topic didn't have a `name` attribute. I had a quick talk with @csweichel last week and we agreed on removing the metric over there.

@csweichel I feel like we had another reason that I can't remember right now. Do you remember if there were more reasons to not expose a metric for this queue?

Comment (csweichel) [reply components/server/src/workspace/messagebus-integration.ts:124]: primarily the fact that it's a queue and not an exchange and we didn't find a good naming scheme.

File: components/server/src/workspace/messagebus-integration.ts [hunk 3]
```diff
@@ -127,10 +129,11 @@ export class MessageBusIntegration extends AbstractMessageBusIntegration {
...
 
-    listenForWorkspaceInstanceUpdates(userId: string | undefined, callback: (ctx: TraceContext, workspaceInstance: WorkspaceInstance) => void): Disposable {
+    async listenForWorkspaceInstanceUpdates(userId: string | undefined, callback: (ctx: TraceContext, workspaceInstance: WorkspaceInstance) => void): Disposable {
```
Comment (akosyakov) [inline components/server/src/workspace/messagebus-integration.ts:132]: this method cannot be async

Comment (akosyakov) [reply components/server/src/workspace/messagebus-integration.ts:132]: the same for other listen methods which return disposable object, it should happen synchronously to be registered fro the current server

Comment (ArthurSens) [reply components/server/src/workspace/messagebus-integration.ts:132]: Very good point, changing it to synchronously would make us lose the ability to use `listener.topic()` which returns the topic name.

I guess this one won't make into tomorrow's release then, I need to think of another way to get the name

Comment (akosyakov) [reply components/server/src/workspace/messagebus-integration.ts:132]: you can do `listener.topic().then(increaseMessagebusTopicReads, () => { /* no-op*/ })`

**Comment by akosyakov** [comment]
btw is topic name meaningful? Last time when I looked directly at messagebus everything had cryptic names maybe we should give better names to topics/queues first?

File: components/server/src/workspace/messagebus-integration.ts (hunk 3 above)
Comment (geropl) [reply components/server/src/workspace/messagebus-integration.ts:132]: We could also try to get rid of the `async` in front of `topic` @ArthurSens ping me if you find time for this today.

**Comment by ArthurSens** [comment]
> btw is topic name meaningful? Last time when I looked directly at messagebus everything had cryptic names maybe we should give better names to topics/queues first?

The result of `listener.topic()` doesn't look good to me, I'd rather hard code a human readable string 😅 
image

**Comment by ArthurSens** [comment]
On commit 7152f82 we have these results:

image

While in 49950cf we have these:
image


@geropl @akosyakov @csweichel which one do you prefer?

**Comment by geropl** [comment]
I like https://github.com/gitpod-io/gitpod/commit/49950cfd358f18b88b3d28bdcaa8c8c1d5fd5451 because it's interesting to thee the number of messages per instance to detect outliers.

If that blows up our scrape-budget I'm fine with the alternative as well.

**Comment by ArthurSens** [comment]
> I like [49950cf](https://github.com/gitpod-io/gitpod/commit/49950cfd358f18b88b3d28bdcaa8c8c1d5fd5451) because it's interesting to thee the number of messages per instance to detect outliers.
> 
> If that blows up our scrape-budget I'm fine with the alternative as well.

That ID is the workspace instance ID? I think it can get very expensive as we scale 😬 Workspace instance ID is the highest cardinality we can get out of Gitpod, besides, how do you plan to use the workspace ID on an alert of a dashboard? I'm afraid it's unactionable data...

**Comment by ArthurSens** [comment]
I'm taking a look at the [TSDB whitepaper](https://www.vldb.org/pvldb/vol8/p1816-teller.pdf) again just to have a better judgment of the costs.

We have 64 bytes for each timeseries header(one timeseries per workspace ID), then 64 bytes for the first sample of each series. Resulting in at least 128 bytes per timeseries.

The next samples' compression rate depends on how much it changes in comparison with the last sample. While testing the metric, I saw some increases of 6~10 to the counters each time I made an interaction with a workspace(create or stop), it's a small increase and will probably lead to less than 1 byte per scrape for each workspace.


image

* Prometheus usually retain metrics in memory for 2~3h before dumping it to disk
* Assuming our scrape interval is 30s, in the worst case scenario we would do 720 scrapes before dumping to disk. 
* Even though workspaces are ephemeral, metrics are not. That means that a metric for a workspace that doesn't exist anymore will keep being exposed until the `server` pod is restarted. Assuming we only restart server once a month (during our monthly updates)

At the end of the month the memory necessary to have prometheus running only to called this metrics would be:
```
(720  x bytes per sample(1) x # of workspace in a month)
+
(128 x # of workspace in a month)
```

Taking a look at our Analytics platform, in march we hosted 209.734 workspace instances, so the results would be:
```
720 x 1 x 209734 = 151008480 bytes
+ 
128 x 209734 = 26845952 bytes

---
151008480 + 26845952 = 177854432 bytes ≅ 169,61 MB 
```

image

I think 170MB is okey for now, but we need to keep in mind that this won't scale well as we grow our community :)

**Comment by ArthurSens** [comment]
If we don't add a workspace instance ID to this metric, the calculation would be:

# of scrapes in 3h(720) x bytes per sample(1) x # of timeseries(3)
720 x 3 = 2160 bytes = 2KB

**Review by meysholdt — APPROVED** [review]
@ArthurSens and I had a quick call and decided that our infra can handle the extra load of having the instance ID included; If we see in self-monitoring at some point in the future that it's too much data we can still turn if off.

**Pull request merged**